### ISSUE TITLE
command to create new definition while proving another one

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean3#484]
In some definitions you want to unfold part of the definition a lot, but another part much less often. For example in the following definition of composition of natural transformations:

``` lean
  protected definition compose [reducible] (η : G ⟹ H) (θ : F ⟹ G) : F ⟹ H :=
  nat_trans.mk
    (λ a, η a ∘ θ a)
    (λ a b f,
      calc
        H f ∘ (η a ∘ θ a) = (H f ∘ η a) ∘ θ a : by rewrite assoc
                      ... = (η b ∘ G f) ∘ θ a : by rewrite naturality
                      ... = η b ∘ (G f ∘ θ a) : by rewrite assoc
                      ... = η b ∘ (θ b ∘ F f) : by rewrite naturality
                      ... = (η b ∘ θ b) ∘ F f : by rewrite assoc)
```

The first argument of `nat_trans.mk` will have to be unfolded a lot, but the second argument is almost always irrelevant. To increase performance, we could create a new lemma which is proves the naturality of this definition, but it would be nice if we could have some block, like `separate ... end` (but probably with a better name) to do this in one definition. 
The name of the created definition does not matter much, we will hardly refer to it explicitly (but it will still be printed sometimes, so perhaps it shouldn't be too unreadable).

It would also be nice if we could specify that the created lemma is `reducible` or `opaque`.

**Comment by leodemoura** [comment]
I was thinking about this feature today. It is feasible to do it for definitions. However, it is a lot more work for theorems because we compile them in parallel, and the parallel compilation is implemented using  a `replace` operation provided by the kernel. This operation allows us to replace an axiom `A` with a theorem `T` IF we have type checked `T` in the same environment we used to type check `A`. This condition makes sure there is no circularity (e.g., we used `A` to prove `T`). 
So, in the current architecture, we cannot say: replace `A` with `auxiliary definitions + T`.

This feature is also useful for definitions that make heavy use of type class resolution (e.g., `decidable` instances). We can add an option for abstracting all non-trivial instances occurring in definitions. This  may be useful for subsingleton types such as `decidable` and `Prop`.

BTW, I'm also fighting with a related problem at #666.

**Comment by fpvandoorn** [comment]
I think restricting this feature just to definitions is fine. We use theorems for statements where we only care that we have an inhabitant, but we don't care about _which_ inhabitant we have. And then this feature is useless, because this feature is only useful when you unfold the definition.

**Comment by leodemoura** [comment]
I added the new feature. The syntax is `abstract ... end`.
We can provide a name: `abstract as foo ... end`
We can provide attributes: `abstract [reducible] ... end`
If we don't provide a name, it adds `_idx` after the name of the declaration.
Here is a list of potential issues:
- The new definition doesn't inherit the `private` and/or `protected` modifier.
- If the definition is inside of a scope with parameters, we don't fix the parameters of the auxiliary definitions inside the section.
- The parameters of the auxiliary definition are not inheriting the binder annotations of the parent definition.

The commit above has some examples.
I also tried the new feature in the HoTT lib: https://github.com/leanprover/lean/commit/62e1be897c235193c5857619e43b3ac7da20caf4

**Comment by fpvandoorn** [comment]
This is great! I've already used this in the files I've been working on last couple of days. I will try to add it to a lot of HoTT definitions soon, and check whether there is a noticeable increase in performance.

Some comments:
- It would be great to also make this available as a command in tactic mode. Currently a workaround is to use `exact abstract begin *tactics* end end`
- Another potential issue: it doesn't inherit the names of the arguments of the parent definition (i.e. the type of the new definition is something like `Pi (x : _) (x_1 : _) (x_2 : _), ...`
- An idea for the future: maybe add an attribute `[abstract n]` where `n` is a (small) natural number. If you give this attribute to an identifier such as `nat_trans.mk`, then the specified argument is _always_ abstracted in definitions. So whenever we write 

``` lean
definition foo (...) : nat_trans ... :=
nat.trans.mk f p
```

this will be interpreted as

``` lean
definition foo (...) : nat_trans ... :=
nat.trans.mk f (abstract p end)
```

If this is implemented, it should be possible to give the `abstract` attribute to more than one argument of a definition.

**Comment by leodemoura** [comment]
The three features are feasible to implement. 
1 and 3 seem useful. Do you care about 2? or is it just a minor annoyance?

**Comment by fpvandoorn** [comment]
`abstract ... end` has another purpose than hiding terms we don't care about, which is the ability to partial unfold a definition. I had a definition which was huge, and which I needed to unfold, but because I had hidden some parts within an `abstract` I could unfold the definition in parts, which was useful.
In this case we do care about the exact definition of the abstracted part, so it might be useful to `print` it, which was exactly how I discovered 2. This does make the definition much less readable.

So I guess it's a bit more than a minor annoyance, but it is in no way critical.
